conan/internal/util/semaphore.py

```diff
@@ -0,0 +1,129 @@
+""" The semaphore module provides inter-process locking mechanisms to ensure Conan commands can
+    run concurrently without conflicts.
+
+    It uses the fasteners library to create and manage locks across multiple processes. Thus, this
+    module is a proxy in case the project need to use a different library in the future.
+"""
+import os
+from datetime import datetime
+from typing import Any
+import inspect
+import traceback
+
+import fasteners
+
+from conan.errors import ConanException
+from conan.api.output import ConanOutput
+from contextlib import contextmanager
+from conan.internal.cache.cache import PkgCache
+
+
+CONAN_SEMAPHORE_FILELOCK = "conan_semaphore.lock"
+
+
+def _filelock_path(conan_api: Any) -> str:
+    """ Get the path to the interprocess file lock.
+
+    :param cache_folder: ConanAPI cache folder path
+    :return: Path to the file lock in Conan cache temporary directory
+    """
+    cache = PkgCache(conan_api.cache_folder, conan_api._api_helpers.global_conf)
+    return os.path.join(cache.filelock_folder, CONAN_SEMAPHORE_FILELOCK)
+
+
+def _raised_by_fasteners(exc: BaseException) -> bool:
+    """
+    Check if the exception was raised by the fasteners library.
+
+    :param exc: The exception to check.
+    :return: True if the exception was raised by fasteners, False otherwise.
+    """
+    for frame, _ in traceback.walk_tb(exc.__traceback__):
+        mod = inspect.getmodule(frame.f_code)
+        if mod and (mod is fasteners or (getattr(mod, '__name__', '').startswith('fasteners'))):
+            return True
+    return False
+
+
+@contextmanager
+def interprocess_lock(conan_api: Any) -> None:
```

Comment:
Passing the full ``ConanAPI`` here to obtain a folder is too much of scope. Better reduce it to the minimum information, if it is a folder, better be a folder.
Note that the ``utils/semaphore.py`` should be a pretty Conan-agnostic set of utilities, not the other way round, that it also creates more import-loops which can be inconvenient.

Reply:
Understood, I'll move it to `${CONAN_HOME}/semaphore/conan_semaphore.lock`. Thank you for explaining.

Reply:
Done, now it will create the file `${CONAN_HOME}/semaphore/conan_semaphore.lock`. Plus, `interprocess_write_lock()` and others only require the Conan home path (much simpler now).